docs/source/CUSTOM-JOB.rst

```diff
@@ -0,0 +1,435 @@
+            image = utils.get_image(fname)
+
+            # preprocess data if necessary
+
+            # send the data to the model
+            results = self.grpc_image(image,
+                                    hvals.get('model_name'),
+                                    hvals.get('model_version'))
+
+            # postprocess results if necessary
+
+            # save the results as an image
+            outpaths = utils.save_numpy_array(results, name=name,
+                                            subdir=subdir, output_dir=tempdir)
+
+            # zip up the file
+            zip_file = utils.zip_files(outpaths, tempdir)
+
+            # upload the zip file to the cloud bucket
+            dest, output_url = self.storage.upload(zip_file)
+
+            # save the results to the redis hash
+            self.update_key(redis_hash, {
+                'status': self.final_status,
+                'output_url': output_url,
+                'output_file_name': dest
+                })
+
+        # return the final status
+        return self.final_status
+
+Finally, the new consumer needs to be registered in the script |consume-redis-events.py| by modifying the function ``get_consumer()`` shown below. Add a new if statement for the new queue type (:data:`queue_name`) and the corresponding consumer.
+
+.. code-block:: python
+
+    def get_consumer(consumer_type, **kwargs):
+        logging.debug('Getting `%s` consumer with args %s.', consumer_type, kwargs)
+        ct = str(consumer_type).lower()
+        if ct == 'image':
+            return redis_consumer.consumers.ImageFileConsumer(**kwargs)
+        if ct == 'zip':
+            return redis_consumer.consumers.ZipFileConsumer(**kwargs)
+        if ct == 'tracking':
+            return redis_consumer.consumers.TrackingConsumer(**kwargs)
+        raise ValueError('Invalid `consumer_type`: "{}"'.format(consumer_type))
+
+.. |consume-redis-events.py| raw:: html
+
+    <tt><a href="https://github.com/vanvalenlab/kiosk-redis-consumer/blob/master/consume-redis-events.py">consume-redis-events.py</a></tt>
+
+Deploying a custom consumer
+---------------------------
+
+The DeepCell Kiosk uses |helm| and |helmfile| to coordinate Docker containers. This allows the :mod:`redis_consumer` to be easily extended by setting up a new docker image with your custom consumer.
+
+1. If you do not already have an account on `Docker Hub <https://hub.docker.com/>`_. Sign in to docker in your local environment using ``docker login``.
+
+2. In an environment of your choice, run ``docker build <image>:<tag>`` and then ``docker push <image>:<tag>``.
+
+3. In the ``/conf/helmfile.d/`` folder in your kiosk environment, add a new helmfile following the convention ``02##.custom-consumer.yaml``. The text for the helmfile can be copied from ``0250.tracking-consumer.yaml`` as shown below. Then make the following changes to customize the helmfile to your consumer.
+
+  * Line 13: Change ``name`` to :data:`consumer_name`
```

Comment:
Using exact line numbers seems fragile.  any future change requires these to be changed significantly.

Reply:
I debated about using specific line numbers. I thought it made the changes a bit easier to navigate if people aren't used to looking at these files. My hope was that by including an example helmfile below this list, it would serve as a static reference which wouldn't need to be updated. Thoughts?

Reply:
I think instead of line numbers we can refer to them as an "object". Like `releases.name` or `releases.values.env.QUEUE`.

This way the reader can see how it gets indented each way, and it remains flexible to any  line changes.

Reply:
This is a good compromise! Change made.